extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/caffeine/CaffeineCache.java

```diff
@@ -50,53 +51,52 @@ public CaffeineCache(CaffeineCacheInfo cacheInfo) {
         cache = builder.buildAsync();
     }
 
-    public Object get(Object key, Callable<Object> valueLoader, long lockTimeout) throws Exception {
-        if (lockTimeout <= 0) {
-            return fromCacheValue(cache.synchronous().get(key, k -> new MappingSupplier(valueLoader).get()));
-        }
-
-        // The lock timeout logic starts here.
-
-        /*
-         * If the current key is not already associated with a value in the Caffeine cache, there's no way to know if the
-         * current thread or another one started the missing value computation. The following variable will be used to
-         * determine whether or not a timeout should be triggered during the computation depending on which thread started it.
-         */
-        boolean[] isCurrentThreadComputation = { false };
+    @Override
+    protected String getName() {
+        return name;
+    }
 
-        CompletableFuture<Object> future = cache.get(key, (k, executor) -> {
-            isCurrentThreadComputation[0] = true;
-            return CompletableFuture.supplyAsync(new MappingSupplier(valueLoader), executor);
-        });
+    @Override
+    public CaffeineCache asCaffeineCache() {
+        return (CaffeineCache) this;
+    }
 
-        if (isCurrentThreadComputation[0]) {
-            // The value is missing and its computation was started from the current thread.
-            // We'll wait for the result no matter how long it takes.
-            return fromCacheValue(future.get());
-        } else {
-            // The value is either already present in the cache or missing and its computation was started from another thread.
-            // We want to retrieve it from the cache within the lock timeout delay.
+    @Override
+    public <T> Uni<T> get(Object key, Callable<T> valueLoader) {
+        if (key == null) {
+            throw new NullPointerException(NULL_KEYS_NOT_SUPPORTED_MSG);
+        }
+        CompletionStage<Object> cacheValueCompletionStage = cache.get(key, unused -> {
```

Comment:
There is one problem with this approach. CompletionStage is a eager type: it starts to compute right away. Uni on the other hand is lazy: it starts to compute when the Uni is effectively asked ot be resolved. 

Is there a solution?

CC @cescoffier

Reply:
I think something along those lines

```
Uni.createFrom().item(() -> //get the completion stage object ...
```

Reply:
I changed this code to:
```
Uni.createFrom().deferred(() -> {
    CompletionStage<Object> cacheValueCompletionStage = cache.get(key, unused -> {
        // omitted code
    });
    return Uni.createFrom().completionStage(cacheValueCompletionStage).map(cacheValue -> {
        return cast(fromCacheValue(cacheValue));
    });
});
```

Reply:
To be continued [here](https://github.com/quarkusio/quarkus/pull/8631#discussion_r557682176).